GREYErrorFormatter.m

```diff
@@ -0,0 +1,195 @@
+static NSString *const kErrorFormatterStackTraceKey         = @"Stack Trace";
+
+/**
+ * Keys used for logging the UI Hierarchy
+ */
+static NSString *const kHierarchyWindowLegendKey                 = @"[Window 1]";
+static NSString *const kHierarchyAcessibilityLegendKey           = @"[AX]";
+static NSString *const kHierarchyUserInteractionEnabledLegendKey = @"[UIE]";
+static NSString *const kHierarchyBackWindowKey                   = @"Back-Most Window";
+static NSString *const kHierarchyAccessibilityKey                = @"Accessibility";
+static NSString *const kHierarchyUserInteractionEnabledKey       = @"User Interaction Enabled";
+static NSString *const kHierarchyLegendKey                       = @"Legend";
+static NSString *const kHierarchyHeaderKey                       = @"UI Hierarchy (ordered by wind"
+                                                                   @"ow level, back to front):\n";
+
+#pragma mark - Private Variables
+
+@interface GREYErrorFormatter ()
+@property(readonly, nonatomic) GREYError *error;
+@end
+
+@implementation GREYErrorFormatter
+
+#pragma mark - Init
+
+- (instancetype)initWithError:(GREYError *)error {
+  self = [super init];
+  if (self) {
+    _error = error;
+  }
+  return self;
+}
+
+#pragma mark - Public Methods
+
+- (NSString *)formattedDescription {
+  if (GREYShouldUseErrorFormatterForError(_error)) {
+    return [self loggerDescriptionForKeys:[self loggerKeys]];
+  }
+  return [GREYObjectFormatter formatDictionary:[_error grey_descriptionDictionary]
+                                        indent:kGREYObjectFormatIndent
+                                     hideEmpty:YES
+                                      keyOrder:nil];
+}
+
+#pragma mark - Private Methods
+
+- (NSSet<NSString *> *)loggerKeys {
+  if ([_error.domain isEqualToString:kGREYInteractionErrorDomain] &&
+       _error.code == kGREYInteractionElementNotFoundErrorCode) {
+    return [[NSSet alloc] initWithArray:@[kErrorFormatterExceptionReasonKey,
+                                          kErrorFormatterRecoverySuggestionKey,
+                                          kErrorFormatterElementMatcherKey,
+                                          kErrorFormatterSearchActionInfoKey,
+                                          kErrorFormatterScreenshotPathsKey,
+                                          kErrorFormatterUnderlyingErrorKey,
+                                          kErrorFormatterUIHierarchyKey,
+                                          kErrorFormatterStackTraceKey]];
+  }
+  GREYFatalAssertWithMessage(false, @"Error Domain and Code Not Yet Supported");
+}
+
+- (NSString *)loggerDescriptionForKeys:(NSSet<NSString *> *)keys {
+  NSMutableArray<NSString *> *logger = [[NSMutableArray alloc] init];
+  
+  if ([keys containsObject:kErrorFormatterExceptionReasonKey]) {
+    NSString *exceptionReason = _error.localizedDescription;
+    [logger addObject:[NSString stringWithFormat:@"\n%@\n", exceptionReason]];
```

Comment:
A lot of format strings like "\n%@\n", "%@\n" are used. I'm wondering if the following helper method will be more readable:

[NSString grey_errorEntryWithSpacing:YES/NO format:@"%@", exceptionReason];

Or we can make it more specific:
[NSString grey_errorEntry:foo] ->`@"\n%@\n", foo`
[NSString grey_oneLinedErrorInfo:info forKey:key] -> `@"\n%@: %@\n", key, info`
[NSString grey_twoLinedErrorInfo:info forKey:key] -> `@"\n%@\n%@\n", key, info`

Reply:
Look at the most recent commit. I think it fixes the issue.